Assets/BossRoom/Scripts/Shared/Net/UnityServices/Lobbies/LobbyContentHeartbeat.cs

```diff
@@ -0,0 +1,146 @@
+        private readonly LobbyUser m_localUser;
+        private int m_awaitingQueryCount = 0;
+        private bool m_shouldPushData = false;
+
+        private readonly UpdateRunner m_SlowUpdate;
+        private IDisposable m_DisposableSubscription;
+        private readonly LobbyAsyncRequests m_LobbyAsyncRequests;
+        private readonly IPublisher<UnityServiceErrorMessage> m_UnityServiceErrorMessagePublisher;
+        private readonly ApplicationController m_ApplicationController;
+
+        [Inject]
+        public LobbyContentHeartbeat(
+            UpdateRunner slowUpdate,
+            LobbyAsyncRequests lobbyAsyncRequests,
+            IPublisher<UnityServiceErrorMessage> unityServiceErrorMessagePublisher,
+            LocalLobby localLobby,
+            LobbyUser localUser,
+            ApplicationController applicationController)
+        {
+            m_SlowUpdate = slowUpdate;
+            m_LobbyAsyncRequests = lobbyAsyncRequests;
+            m_UnityServiceErrorMessagePublisher = unityServiceErrorMessagePublisher;
+            m_localLobby = localLobby;
+            m_localUser = localUser;
+            m_ApplicationController = applicationController;
+        }
+
+        public void BeginTracking()
+        {
+            m_SlowUpdate.Subscribe(OnUpdate, 1.5f);
+
+            m_localLobby.onChanged += OnLocalLobbyChanged;
+            m_shouldPushData = true; // Ensure the initial presence of a new player is pushed to the lobby; otherwise, when a non-host joins, the LocalLobby never receives their data until they push something new.
+        }
+
+        public void EndTracking()
+        {
+            m_shouldPushData = false;
+            m_SlowUpdate.Unsubscribe(OnUpdate);
+            //disposing of subscription automatically removes the subscription from MessageChannel
+            m_DisposableSubscription?.Dispose();
+            if (m_localLobby != null)
+            {
+                m_localLobby.onChanged -= OnLocalLobbyChanged;
+            }
+        }
+
+        private void OnLocalLobbyChanged(LocalLobby changed)
+        {
+            if (string.IsNullOrEmpty(changed.LobbyID)) // When the player leaves, their LocalLobby is cleared out but maintained.
+            {
+                EndTracking();
+            }
+
+            m_shouldPushData = true;
+        }
+
+
+        /// <summary>
+        /// If there have been any data changes since the last update, push them to Lobby. Regardless, pull for the most recent data.
+        /// (Unless we're already awaiting a query, in which case continue waiting.)
+        /// </summary>
+        private void OnUpdate(float dt)
+        {
+            if (m_awaitingQueryCount > 0)
+            {
+                return;
+            }
+
+            if (m_localUser.IsHost)
+            {
+                m_LobbyAsyncRequests.DoLobbyHeartbeat(dt);
+            }
+
+            if (m_shouldPushData)
+            {
+                PushDataToLobby();

```

Comment:
wait, why do we do this in an update loop? Why isn't this directly called?

Reply:
It's done in an update loop, yes, but not in a Unity update loop - it's using Update runner and the function is actually OnUpdate(float dt)

It's basically an infrequent timer-based data push attempt done every 1.5 seconds (and the data is sent only if it's actually dirty). And if you're the client - it just tries to refresh the lobby. 

I have an idea of replacing this with message channel, now that we're using it extensively.

Reply:
Sorry, my question wasn't clear, why not update lobbies on click when we change them? Why polling for data changes? Why not just send changes as soon as they are produced?

Reply:
In other words, why not call UpdateLobbyDataAsync directly in OnChange instead of having the m_ShouldPushData set to true and then having to wait for the next update?

Reply:
I'm guessing cause you don't want to spam the lobby service and trigger rate limiting? I'd really write this down in a comment somewhere, users reading this for the first time will miss this.
(This should really be feedback to the SDK team, this isn't great that users have to write this themselves)